src/main/java/org/opentripplanner/routing/algorithm/raptoradapter/transit/mappers/McCostParamsMapper.java

```diff
@@ -25,8 +32,26 @@ public static McCostParams map(RoutingRequest request) {
 
     builder.wheelchairAccessibility(request.wheelchairAccessibility);
 
-    builder.unpreferredRoutes(request.unpreferredRoutes.stream().collect(Collectors.toSet()));
-    builder.unpreferredCost(request.unpreferredRouteCost);
+    final BitSet unpreferredPatterns = new BitSet();
+    final Set<FeedScopedId> unpreferredRoutes = request.getUnpreferredRoutes();
+    final Set<FeedScopedId> unpreferredAgencies = request.getUnpreferredAgencies();
+
+    if (!unpreferredRoutes.isEmpty() || !unpreferredAgencies.isEmpty()) {
+      for (var pattern : patternIndex) {
+        if (
+          pattern != null &&
+          (
+            unpreferredRoutes.contains(pattern.route().getId()) ||
+            unpreferredAgencies.contains(pattern.route().getAgency().getId())
+          )
+        ) {
+          unpreferredPatterns.set(pattern.patternIndex());
+        }
+      }
+    }
+
+    builder.unpreferredPatterns(unpreferredPatterns);
+    builder.unpreferredCost(request.unpreferredCost);
```

Comment:
If there are no unpreferred patterns we can skip adding them to the builder:

```suggestion
      builder.unpreferredPatterns(unpreferredPatterns);
      builder.unpreferredCost(request.unpreferredCost);
    }
```

Reply:
Done